src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,288 @@
+      | Power
+      | Divide => Some(Typ.Float)
+      | LessThan
+      | LessThanOrEqual
+      | GreaterThan
+      | GreaterThanOrEqual
+      | Equals
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinStringOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, String) && Typ.eq(ty2, String)) {
+      switch (op) {
+      | Concat => Some(Typ.String)
+      | Equals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | ListLit(_, _, ty, _) => Some(List(ty))
+  | Cons(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty2) {
+    | List(Unknown(Internal)) => Some(Typ.List(ty1))
+    | List(ty3) when Typ.eq(ty3, ty1) => Some(ty2)
+    | _ => None
+    };
+  | ListConcat(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty1, ty2) {
+    | (List(Unknown(Internal)), _)
+    | (_, List(Unknown(Internal))) => Some(Typ.List(Unknown(Internal)))
+    | (List(ty1), List(ty2)) when Typ.eq(ty1, ty2) => Some(Typ.List(ty1))
+    | _ => None
+    };
+  | Tuple(dhs) =>
+    let+ typ_list =
+      dhs |> List.map(typ_of_dhexp(ctx, m)) |> OptUtil.sequence;
+    Typ.Prod(typ_list);
+  | Prj(dh, i) =>
+    let* ty = typ_of_dhexp(ctx, m, dh);
+    switch (ty) {
+    | Prod(l) when List.length(l) != 0 => Some(List.nth(l, i))
+    | _ => None
+    };
+  | Constructor(_, typ) => Some(typ)
+  | ConsistentCase(Case(d_scrut, d_rules, _)) =>
+    let* ty' = typ_of_dhexp(ctx, m, d_scrut);
+    let* typ_cases: list(Typ.t) =
+      d_rules
+      |> List.map(rule_prj)
+      |> List.map(((dhp, de)) => {
+           typ_of_dhexp(dhpat_extend_ctx(dhp, ty', ctx), m, de)
+         })
+      |> OptUtil.sequence;
+    Typ.join_all(~empty=Unknown(Internal), ctx, typ_cases);
+  | Cast(d, ty1, ty2) =>
+    let* _ = Typ.join(~fix=true, ctx, ty1, ty2);
+    let* tyd = typ_of_dhexp(ctx, m, d);
+    Typ.eq(tyd, ty1) ? Some(ty2) : None;
+  | FailedCast(d, ty1, ty2) =>
+    if (ground(ty1) && ground(ty2) && !Typ.eq(ty1, ty2)) {
+      let* tyd = typ_of_dhexp(ctx, m, d);
+      Typ.eq(tyd, ty1) ? Some(ty2) : None;
+    } else {
+      None;
+    }
+  | InvalidOperation(_) => None
+  | IfThenElse(ConsistentIf, d_scrut, d1, d2) =>
+    let* ty = typ_of_dhexp(ctx, m, d_scrut);
+    if (Typ.eq(ty, Bool)) {
+      let* ty1 = typ_of_dhexp(ctx, m, d1);
+      let* ty2 = typ_of_dhexp(ctx, m, d2);
+      Typ.join_all(~empty=Unknown(Internal), ctx, [ty1, ty2]);
+    } else {
+      None;
+    };
+  | IfThenElse(InconsistentIf, d_scrut, d1, d2) =>
+    let* ty = typ_of_dhexp(ctx, m, d_scrut);
+    if (Typ.eq(ty, Bool)) {
+      let* _ = typ_of_dhexp(ctx, m, d1);
+      let+ _ = typ_of_dhexp(ctx, m, d2);
+      Typ.Unknown(Internal);
+    } else {
+      None;
+    };
+  };
+};
+
+let property_test = (uexp_typ: Typ.t, dhexp: DHExp.t, m: Statics.Map.t): bool => {
```

Comment:
it would be good to surface these tests directly around the elaboration function to the user, so we can rapidly fix any issues that come up in elaboration